Improve Build Docs and Agile Coretime - Coreify The Build Section (#5811)
w3f/polkadot-wiki: docs/build/build-guide.md

slug: ../build-guide
---

import Tabs from "@theme/Tabs"; import TabItem from "@theme/TabItem";

{{ polkadot: Polkadot :polkadot }}{{ kusama: Kusama :kusama }} is a blockchain protocol and network
that provides **shared security** amongst all connected parachains and allowing all connected tasks

> to **interoperate** by using [XCM](../learn/learn-xcm.md).

Reviewer: ```suggestion
{{ polkadot: Polkadot :polkadot }}{{ kusama: Kusama :kusama }} is a blockchain network protocol
that provides **shared security** among all connected parachains and allowing all connected tasks
to **interoperate** by using [XCM](../learn/learn-xcm.md).
```

The time it takes to develop and launch a new chain has dropped significantly with the tools
Polkadot provides. While before, it would take years to launch a new chain, now it may only take

> weeks or even days, depending on your goals.

Reviewer: Adding link to Polkadot SDK. Dropping the line that says it takes years to launch a chain :) Probably was written several years ago. Now that statement carries little weight. Instead, we can focus on showing how quickly we can deploy chains through Agile coretime.

```suggestion
With the [Polkadot SDK](https://github.com/paritytech/polkadot-sdk) and [Agile Coretime](../learn/learn-agile-coretime.md), the time it takes to develop and launch a new chain has dropped significantly. Depending on your goals, it may only take weeks or even days.
```

This starters guide will walk you through the steps you can take today to get started building your
vision with {{ polkadot: Polkadot :polkadot }}{{ kusama: Kusama :kusama }}. It will also point to
resources you can use to start building immediately.

:::note For Developers!

Please note that the following documentation is more geared towards developers. If you are looking

> for more general knowledge related to Polkadot, be sure to check otu the

Reviewer: ```suggestion
for more general knowledge related to Polkadot, be sure to check out the
```

[Learn](../learn/learn-index) section.

> Looking to jump into a guide? Check out our starter developer guides for creating a minimal

Reviewer: Where are the starter developer guides? Could you please link them here?

Author: I had originally planned to include the section on getting started with coretime, which this PR bootstraps (but does not add all the content for, yet).  I can remove it until it is all in place.

Reviewer: Got it

blockchain and deploying it on a testnet.

:::

This build guide covers three different areas, taking a **top-down** approach from protocol
development to user-facing applications:

<!-- prettier-ignore -->
<!-- no toc -->

> 1. [Protocol Development - Developing parachains/blockchains](#building-parachains-tasks)

Reviewer: Shouldn't this be called runtime development?

Author: I suppose, although it includes the full stack in many cases, i.e., from running a collator, consensus mechanisms (for example, minimal and solo-chain containing different block authoring mechanisms, both of which are reflected in `node/`) I can change it if it doesn't make sense.

Reviewer: As it was pointing to #building-parachains-tasks section, I thought it makes sense to call it runtime development as the parachain protocol will not be touched.

Author: yeah that makes sense - will refactor accordingly

{{ polkadot: Polkadot :polkadot }}{{ kusama: Kusama :kusama }} and its networks in order to gain
context about the application you could make.

[Take a look at the various development network options here.](./build-network-overview.md)

## Building Parachains

{{ polkadot: Polkadot :polkadot }}{{ kusama: Kusama :kusama }} is canonically referred to as the
Relay Chain. It is also considered a _layer zero_ protocol, as it enables the interoperability and
shared security of multiple [parachains](../learn/learn-parachains.md), which are _layer one_
protocols. Parachains currently connect to a relay chain using the
[Parachains Protocol](../learn/learn-parachains-protocol.md). More elaborate (or simpler) tasks
could be constructed in the future.

![build 1](../assets/build-1.png)

:::info

Throughout this document, you may encounter the term **runtime** or **STF (State Transition
Function)**. Both refer to the same concept, as they define how a particular system, i.e., a
blockchain, should deal with state changes externally and internally. Both of these terms are used
extensively in {{ polkadot: Polkadot :polkadot }}{{ kusama: Kusama :kusama }} and Substrate.

:::

> The Polkadot SDK opens possibilities to construct complex **runtime**, or **STF** (state transition

Reviewer: Here, we are comparing parachains with smart contracts. Hence, including the term to retain the context. 

```suggestion
Parachains built through the Polkadot SDK, open possibilities to construct complex **runtime**, or **STF** (state transition
```

Currently, the most streamlined way of utilizing Substrate is through
[FRAME](https://paritytech.github.io/polkadot-sdk/master/polkadot_sdk_docs/polkadot_sdk/frame_runtime/index.html),
which conveniently allows for a runtime/STF to be generated from a set of modules (called
[pallets](https://paritytech.github.io/polkadot-sdk/master/polkadot_sdk_docs/polkadot_sdk/frame_runtime/index.html#pallets)).
Runtimes in Substrate are built using [WebAssembly](../learn/learn-wasm.md) (Wasm), and represent
the state transition function for a network. FRAME provides a framework for
[pallets](https://paritytech.github.io/polkadot-sdk/master/polkadot_sdk_docs/polkadot_sdk/frame_runtime/index.html#pallets),
to construct a runtime/STF and define how your task is supposed to behave. Ranging from
[identity](https://paritytech.github.io/substrate/master/pallet_identity/index.html) to
[smart contracts](https://paritytech.github.io/substrate/master/pallet_contracts/index.html),
pallets can be quite extensive in providing on-chain functionality.

Even though FRAME is heavily used, it is not the only way to create a valid runtime/STF using

> Substrate. Substrate can be used to create new paradigms and abstractions, one such example is the
> [Open Runtime Module Library (ORML)](https://github.com/open-web3-stack/open-runtime-module-library),
> which is another way creating and using runtime modules.

Reviewer: ```suggestion
Substrate. Substrate can be used to create new paradigms and abstractions. One such example is the
[Open Runtime Module Library (ORML)](https://github.com/open-web3-stack/open-runtime-module-library),
which is another way of creating and using runtime modules.
```

#### Building Parachains with Cumulus

> [Cumulus](https://paritytech.github.io/polkadot-sdk/master/polkadot_sdk_docs/polkadot_sdk/cumulus/index.html)
> transforms FRAME-based runtimes into Polkadot-compatible parachain runtimes, and Substrate-based
> nodes into Polkadot/Parachain-compatible nodes. In other words, Cumulus can enable a regular runtime
> to meet the requirements to have its state verified by Polkadot and take part in shared security.

Reviewer: Adding more info on the specifics of Cumulus. Ref: https://substrate.stackexchange.com/questions/12/what-is-cumulus

It does not **transform** Substrate-based node into a Polkadot/Parachain node per se. It moves the blockchain runtime onto the collator nodes who produce blocks and send the PoVs to relay chain, instead of finding consensus within their own network.

```suggestion
[Cumulus](https://paritytech.github.io/polkadot-sdk/master/polkadot_sdk_docs/polkadot_sdk/cumulus/index.html) is a set of tools that allows you to convert a blockchain developed using Substrate and FRAME into a Polkadot-compatible Parachain. More specifically, it provides libraries for all of the necessary parts of the Polkadot protocol necessary for Parachains to work, for example:

- Creating new parachain blocks via Collators
- Listening to the relay chain for updates
- Synchronizing upgrades between the parachain and relay chain

```

Parachain Protocol.

:::info

_Cumulus clouds are shaped like dots; together, they form an intricate, beautiful and functional
system._

:::
[Cumulus](https://paritytech.github.io/polkadot-sdk/master/polkadot_sdk_docs/polkadot_sdk/cumulus/index.html)
transforms FRAME-based runtimes into Polkadot-compatible parachain runtimes, and Substrate-based
nodes into Polkadot/Parachain-compatible nodes. In other words, Cumulus can enable a regular runtime
to meet the requirements to have its state verified by Polkadot and take part in shared security.

For most developers, the best place to start is to get familiar with Substrate independently,
followed by FRAME, with Cumulus as the final step to understanding the entire parachain building
process. This way, one can view how various paradigms are applied and decide on integrating or
utilizing Substrate for their particular use case.

Please see the [parachain development guide](build-parachains.md) for how to get started on building
a parachain or parathread.

> #### Parachains Benefits

Reviewer: Why are we getting rid of this section?

Author: It seemed repetitive; the parachain section in Learn should aim to "sell" people on the benefits of parachains. I was aiming to make this more purely focused on building as well as more concise.

Reviewer: Sounds good. Is there a place on the Wiki where this content can be moved to?

Author: I'll look to adding it to the Learn section, most likely under "Why Parachains?" https://wiki.polkadot.network/docs/learn-parachains#why-parachains

WDYT?

Reviewer: Sounds good. Thanks!

[pallet_contracts](https://paritytech.github.io/substrate/master/pallet_contracts/index.html)).
Pallets require more engineering and thought, as they can directly affect the chain's state.

For a more comprehensive (and maintained) comparison, be sure to check out the
[comparison from the Polkadot SDK documentation](https://paritytech.github.io/polkadot-sdk/master/polkadot_sdk_docs/reference_docs/runtime_vs_smart_contract/index.html).

:::

### ink! and EVM-based Smart Contracts

At a high level, a _smart contract_ is simply some code that exists at an address on a chain and is

> callable by external actors. Whether it's [EVM-based](https://polkadot-evm.github.io/frontier/), or
> written using [ink!](https://use.ink/), smart contracts sandboxed, executable programs that live
> on-chain.

Reviewer: ```suggestion
callable by external actors. Whether it's [EVM-based](https://polkadot-evm.github.io/frontier/) or
written using [ink!](https://use.ink/), smart contracts are sandboxed, executable programs that live
on-chain.
```
